okhttp/src/main/java/okhttp3/Request.kt

```diff
@@ -0,0 +1,236 @@
+
+  fun url(): HttpUrl = url
+
+  fun method(): String = method
+
+  fun headers(): Headers = headers
+
+  fun header(name: String): String? = headers[name]
+
+  fun headers(name: String): List<String> = headers.values(name)
+
+  fun body(): RequestBody? = body
+
+  /**
+   * Returns the tag attached with `Object.class` as a key, or `null` if no tag is attached with
+   * that key.
+   *
+   * Prior to OkHttp 3.11, this method never returned `null` if no tag was attached. Instead it
+   * returned either this request, or the request upon which this request was derived with
+   * [newBuilder].
+   */
+  fun tag(): Any? = tag(Any::class.java)
+
+  /**
+   * Returns the tag attached with [type] as a key, or `null` if no tag is attached with that
+   * key.
+   */
+  fun <T> tag(type: Class<out T>): T? = type.cast(tags[type])
+
+  fun newBuilder(): Builder = Builder(this)
+
+  /**
+   * Returns the cache control directives for this response. This is never `null`, even if this
+   * response contains no `Cache-Control` header.
+   */
+  fun cacheControl(): CacheControl = cacheControl
+
+  override fun toString(): String = "Request{method=$method, url=$url, tags=$tags}"
+
+  open class Builder {
+    internal var url: HttpUrl? = null
+    internal var method: String
+    internal var headers: Headers.Builder
+    internal var body: RequestBody? = null
+
+    /** A mutable map of tags, or an immutable empty map if we don't have any. */
+    internal var tags: MutableMap<Class<*>, Any> = mutableMapOf()
+
+    constructor() {
+      this.method = "GET"
+      this.headers = Headers.Builder()
+    }
+
+    internal constructor(request: Request) {
+      this.url = request.url
+      this.method = request.method
+      this.body = request.body
+      this.tags = if (request.tags.isEmpty())
+        mutableMapOf()
+      else
+        request.tags.toMutableMap()
+      this.headers = request.headers.newBuilder()
+    }
+
+    open fun url(url: HttpUrl): Builder = apply {
+      this.url = url
+    }
+
+    /**
+     * Sets the URL target of this request.
+     *
+     * @throws IllegalArgumentException if [url] is not a valid HTTP or HTTPS URL. Avoid this
+     * exception by calling [HttpUrl.parse]; it returns `null` for invalid URLs.
+     */
+    open fun url(url: String): Builder {
+      // Silently replace web socket URLs with HTTP URLs.
+      val finalUrl: String = when {
+        url.regionMatches(0, "ws:", 0, 3, ignoreCase = true) -> {
+          "http:${url.substring(3)}"
+        }
+        url.regionMatches(0, "wss:", 0, 4, ignoreCase = true) -> {
+          "https:${url.substring(4)}"
+        }
+        else -> url
+      }
+
+      return url(HttpUrl[finalUrl])
+    }
+
+    /**
+     * Sets the URL target of this request.
+     *
+     * @throws IllegalArgumentException if the scheme of [url] is not `http` or `https`.
+     */
+    // Code review TODO - this never checked the scheme before like the doc says. Should it?
```

Comment:
I think it’ll explode on HttpUrl.get()

Reply:
kk, will remove this line and leave as is

Reply:
cf36f06